shared/src/main/scala/io/kaitai/struct/ClassCompiler.scala

```diff
@@ -298,6 +352,39 @@ class ClassCompiler(
     }
   }
 
+  def compileFetchInstances(attrs: List[AttrLikeSpec]): Unit = {
+    attrs.foreach { (attr) =>
+      lang.attrFetchInstances(attr, attr.id)
+    }
+  }
+
+  def compileSetInstanceWriteFlags(instances: Map[InstanceIdentifier, InstanceSpec]) = {
+    instances.foreach { case (instName, instSpec) =>
+      instSpec match {
+        case _: ParseInstanceSpec =>
+          lang.instanceSetWriteFlag(instName)
+        case _: ValueInstanceSpec => // do nothing
+      }
+    }
+  }
+
+  def compileSeqWrite(seq: List[AttrSpec], defEndian: Option[FixedEndian]) = {
+    var wasUnaligned = false
+    seq.foreach { (attr) =>
+      val nowUnaligned = isUnalignedBits(attr.dataType)
+      if (wasUnaligned && !nowUnaligned)
+        lang.alignToByte(lang.normalIO)
+      lang.attrWrite(attr, attr.id, defEndian)
+      wasUnaligned = nowUnaligned
+    }
```

Comment:
Let's try to avoid using `var` and imperative. `foldLeft` seems to do the trick here, cleanly conveying the semantics and allowing e.g. advanced parallelization of such loops:

```suggestion
    seq.foldLeft(false) { case (wasUnaligned, attr) =>
      val nowUnaligned = isUnalignedBits(attr.dataType)
      if (wasUnaligned && !nowUnaligned)
        lang.alignToByte(lang.normalIO)
      lang.attrWrite(attr, attr.id, defEndian)
      nowUnaligned
    }
```

Reply:
Actually, one of the major changes in Java and Python runtime libraries related to serialization was that all `alignToByte` logic is handled in the runtime libraries themselves and the `alignToByte()` call generation is disabled in the compiler in both Java and Python:

https://github.com/kaitai-io/kaitai_struct_compiler/blob/2eca3de2c4044b74226d6fcd2f3641a65febd9ab/shared/src/main/scala/io/kaitai/struct/languages/JavaCompiler.scala#L497-L500

https://github.com/kaitai-io/kaitai_struct_compiler/blob/2eca3de2c4044b74226d6fcd2f3641a65febd9ab/shared/src/main/scala/io/kaitai/struct/languages/PythonCompiler.scala#L486-L489

(and now I see a small mistake in the comment in `PythonCompiler` - technically it should be "the compiler does not need to output <code><del>alignToByte</del>()</code> <code><ins><strong>align_to_byte</strong></ins>()</code>" per Python naming)

---

So I guess the piece of code you're reviewing can be simplified to just `lang.attrWrite(attr, attr.id, defEndian)`.